include/rocksdb/configurable.h

```diff
@@ -0,0 +1,491 @@
+   * itself or the DBOptions or the ColumnFamilyOptions (unlike
+   * ValidateOptions). After calling this method, the
+   * DBOptions/ColumnFamilyOptions should be valid. a non-ok Status will be
+   * returned.
+   */
+  Status SanitizeOptions();
+  Status SanitizeOptions(DBOptions& db_opts);
+  Status SanitizeOptions(DBOptions& db_opts, ColumnFamilyOptions& cf_opts);
+
+  /**
+   * Validates the local options are valid for the specified DB Options
+   * and ColumnFamilyOptions. If the options are not valid,
+   * a non-ok Status will be returned.  This method cannot change its input
+   * arguments or the object itself.
+   */
+  Status ValidateOptions() const;
+  Status ValidateOptions(const DBOptions&) const;
+  Status ValidateOptions(const DBOptions& db_opts,
+                         const ColumnFamilyOptions& cf_opts) const;
+
+  /**
+   * Prints the configuration object settings to the input Logger
+   */
+  void Dump(Logger* log, const std::string& indent,
+            uint32_t mode = OptionStringMode::kOptionNone) const {
+    DumpOptions(log, indent, mode);
+  }
+  void Dump(Logger* log, uint32_t mode = OptionStringMode::kOptionNone) const {
+    Dump(log, "              ", mode);
+  }
+
+  /**
+   * Returns the raw pointer of the named options that is used by this
+   * object, or nullptr if this function is not supported.
+   * Since the return value is a raw pointer, the object owns the
+   * pointer and the caller should not delete the pointer.
+   *
+   * Note that changing the underlying options while the object
+   * is currently used by any open DB is undefined behavior.
+   * Developers should use DB::SetOption() instead to dynamically change
+   * options while the DB is open.
+   */
+  template <typename T>
+  const T* GetOptions(const std::string& name) const {
+    return reinterpret_cast<const T*>(GetOptionsPtr(name));
+  }
+  template <typename T>
+  T* GetOptions(const std::string& name) {
+    return reinterpret_cast<T*>(const_cast<void*>(GetOptionsPtr(name)));
+  }
+
+#ifndef ROCKSDB_LITE
+  /**
+   * Configures the options for this extension based on the input parameters.
+   * Returns an OK status if configuration was successful.
+   * Parameters:
+   *   opt_map                 The name/value pair map of the options to set
+   *   opt_str                 String of name/value pairs of the options to set
+   *   input_strings_escaped   True if the strings are escaped (old-style?)
+   *   ignore_unusued_options  If true and there are any unused options,
+   *                           they will be ignored and OK will be returned
+   *   unused_opts             The set of any unused option names from the map
+   */
+  Status ConfigureFromMap(const DBOptions&,
+                          const std::unordered_map<std::string, std::string>&,
+                          bool input_strings_escaped = false,
+                          bool ignore_unused_options = false);
+  Status ConfigureFromMap(const DBOptions&,
```

Comment:
I think we should avoid default arguments for both `ConfigureFromMap` and `ConfigureFromString`.
Compilers may perform implicit conversion from pointer to boolean types (https://en.cppreference.com/w/cpp/language/implicit_conversion) rather than report error as we expect.
An example is as follows.
When you accidentally call `ConfigureFromMap(..., unused)` in which `unused` is a pointer to `std::set`, the compiler will cast `unused` to a boolean and call `ConfigureFromMap(const DBOptions&, const std::unordered_map<std::string, std::string>&, bool, bool)` instead of complaining about type mismatch.